storage/oasis/nodeapi/file/node.go

```diff
@@ -0,0 +1,293 @@
+package file
+
+import (
+	"bytes"
+	"context"
+	"encoding/gob"
+	"encoding/json"
+	"fmt"
+
+	"github.com/akrylysov/pogreb"
+	beacon "github.com/oasisprotocol/oasis-core/go/beacon/api"
+	coreCommon "github.com/oasisprotocol/oasis-core/go/common"
+	consensus "github.com/oasisprotocol/oasis-core/go/consensus/api"
+	genesis "github.com/oasisprotocol/oasis-core/go/genesis/api"
+
+	"github.com/oasisprotocol/oasis-indexer/storage/oasis/nodeapi"
+	"github.com/oasisprotocol/oasis-indexer/storage/oasis/nodeapi/damask"
+)
+
+// FileConsensusApiLite provides access to the consensus API of an Oasis node.
+// Since FileConsensusApiLite is backed by a file containing the cached responses
+// to `ConsensusApiLite` calls, this data is inherently compatible with the
+// current indexer and can thus handle heights from both Cobalt/Damask.
+type FileConsensusApiLite struct {
+	db        pogreb.DB
+	damaskApi *damask.DamaskConsensusApiLite
+}
+
+type ConsensusApiMethod func() (interface{}, error)
+
+var _ nodeapi.ConsensusApiLite = (*FileConsensusApiLite)(nil)
+
+func NewFileConsensusApiLite(filename string, client *consensus.ClientBackend) (*FileConsensusApiLite, error) {
+	db, err := pogreb.Open(filename, &pogreb.Options{BackgroundSyncInterval: -1})
+	if err != nil {
+		return nil, err
+	}
+	var damaskApi *damask.DamaskConsensusApiLite
+	if client != nil {
+		damaskApi = damask.NewDamaskConsensusApiLite(*client)
+	}
+	return &FileConsensusApiLite{
+		db:        *db,
+		damaskApi: damaskApi,
+	}, nil
+}
+
+func generateCacheKey(methodName string, params ...interface{}) ([]byte, error) {
+	var buf bytes.Buffer
+	enc := gob.NewEncoder(&buf)
+	err := enc.Encode(methodName)
+	if err != nil {
+		return nil, err
+	}
+	for _, p := range params {
+		err = enc.Encode(p)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return buf.Bytes(), nil
+}
+
+func (c *FileConsensusApiLite) get(key []byte, result interface{}) error {
+	res, err := c.db.Get(key)
+	if err != nil {
+		return err
+	}
+	return json.Unmarshal(res, result)
+}
+
+func (c *FileConsensusApiLite) put(key []byte, val interface{}) error {
+	valBytes, err := json.Marshal(val)
+	if err != nil {
+		fmt.Printf("file backend: error encoding value for put " + err.Error())
+		return err
+	}
+	return c.db.Put(key, valBytes)
```

Comment:
Can you check if `pogreb` is thread-safe? If not, let's hide `Put()` and `Get()` and `Has()` any other relevant methods behind a mutex.

Reply:
Yep it's goroutine-safe